lib/chef/resource.rb

```diff
@@ -1012,6 +1010,75 @@ def default_action(action_name=NULL_ARG)
       def default_action=(action_name)
         default_action action_name
       end
+
+      #
+      # Define an action on this resource.
+      #
+      # The action is defined as a *recipe* block that will be compiled and then
+      # converged when the action is taken (when Resource is converged).  The recipe
+      # has access to the resource's attributes and methods, as well as the Chef
+      # recipe DSL.
+      #
+      # Resources in the action recipe may notify and subscribe to other resources
+      # within the action recipe, but cannot notify or subscribe to resources
+      # in the main Chef run.
+      #
+      # Resource actions are *inheritable*: if resource A defines `action :create`
+      # and B is a subclass of A, B gets all of A's actions.  Additionally,
+      # resource B can define `action :create` and call `super()` to invoke A's
+      # action code.
+      #
+      # @param name [Symbol] The action name to define.
+      # @param recipe_block The recipe to run when the action is taken. This block
+      #   takes no parameters, and will be evaluated in a new context containing:
+      #
+      #   - The resource's public and protected methods (including attributes)
+      #   - The Chef Recipe DSL (file, etc.)
+      #   - super() referring to the parent version of the action (if any)
+      #
+      # @return The Action class implementing the action
+      #
+      def action(action, &recipe_block)
+        action = action.to_sym
+        create_action_provider_class.action(action, &recipe_block)
+        self.allowed_actions += [ action ]
+        default_action action if default_action == :nothing
+      end
+
+      #
+      # The created action provider class for this resource, or nil if it has
+      # not been / does not need to be created.
+      #
+      # @api private
+      def action_provider_class
+        @action_provider_class ||
+          # If the superclass needed one, then we need one as well.
```

Comment:
This comment doesn't explain _why_.  When would we not need a `action_provider_class`?